pdudaemon/drivers/shelly_gen2.py

```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/python3
+
+#  Shelly Gen2+ driver to control second generation and later SmartPlugs, etc.
+#  made by Shelly.
+#
+#  This does not support the secure requests through the authentication process
+#  described in https://shelly-api-docs.shelly.cloud/gen2/General/Authentication.
+#  Thus, this must be turned off in your Shelly device.
+#
+#  Based on the official Gen 2+ Device API documentation:
+#  https://shelly-api-docs.shelly.cloud/gen2/General/RPCProtocol
+#
+#  Copyright 2025 Jonas Jelonek <jelonek.jonas@gmail.com>
+#
+#  This program is free software; you can redistribute it and/or modify
+#  it under the terms of the GNU General Public License as published by
+#  the Free Software Foundation; either version 2 of the License, or
+#  (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with this program; if not, write to the Free Software
+#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+#  MA 02110-1301, USA.
+
+import logging
+import os
+from pdudaemon.drivers.driver import PDUDriver
+import requests
+
+log = logging.getLogger("pdud.drivers." + os.path.basename(__file__))
+
+class ShellyGen2(PDUDriver):
+    def __init__(self, hostname, settings):
+        self.hostname = hostname
+        self.settings = settings
+
+        super(ShellyGen2, self).__init__()
+
+    def jsonrpc_call(self, method, params):
+        r = requests.post(
+            f"http://{self.hostname}/rpc",
+            json = {
+                "jsonrpc":"2.0",
+                "id": 1,
+                "method": method,
+                "params": params,
+            }
+        )
+
+        r.raise_for_status()
+        return r.json()["result"]
+
+    def port_get(self, port_number):
+        r = self.jsonrpc_call("Switch.GetStatus", {
+            "id": int(port_number) - 1
+        })
+        return r["output"]
+
+    def port_interaction(self, command, port_number):
+        if command == "get":
+            return self.port_get(port_number)
+
+        self.jsonrpc_call("Switch.Set", {
+            "id": int(port_number) - 1,
```

Comment:
ooi is the port_number on the API/device itself 1-based, but the api 0-based? 

You probably should check if it's a valid port number as well. Most drivers will know the number of ports they have

Reply:
From pdudaemon i don't think anything is enforced though most drivers are 1-based i think. At least one  is 0 based, judging by a quick grep.. 

My personal recommendation would be what is easiest for the user to avoid confusion. Which in my view is following whatever the  device labelling and/or its user interfaces do.   If those are inconsistent, I'd probably go for the physical labeling.